### PR TITLE
Pythia Z cross section measurement

### PR DESCRIPTION
**Pull request by r-preston**
@mvesteri Pythia 8.3 should be accessible at `/home/epp/phuswg/pythia8302`

### CODE REVIEW

**Comment by mvesteri** [comment]
👍 @r-preston . One quick question. Can you see the following file?
```
/cvmfs/sft.cern.ch/lcg/views/LCG_98/x86_64-centos7-gcc8-opt/lib/libpythia8.so
```

File: z-theory/generate_z_events.cpp [hunk 1]
```diff
@@ -1,61 +1,30 @@
...
-  // Create Pythia instance and set it up to generate hard QCD processes
-  // above pTHat = 20 GeV for pp collisions at 14 TeV.
+int main()
+{
   Pythia pythia;
-  pythia.readString("HardQCD:all = on");
+  pythia.readString("WeakSingleBoson:all = on");
   pythia.readString("PhaseSpace:pTHatMin = 20.");
-  pythia.readString("Beams:eCM = 14000.");
+  pythia.readString("Beams:eCM = 7000.");
```
Comment (mvesteri) [inline z-theory/generate_z_events.cpp]: Suggest 5000 [GeV] here, to go with the dataset that we are analysing.

File: z-theory/generate_z_events.cpp [hunk 2]
```diff
@@ -1,61 +1,30 @@
...
   pythia.init();
 
-  // Set up the ROOT TFile and TTree.
-  TFile *file = TFile::Open("pytree.root","recreate");
-  Event *event = &pythia.event;
-  TTree *T = new TTree("T","ev1 Tree");
-  T->Branch("event",&event);
+  
+  Event& event = pythia.event;
```
Comment (mvesteri) [inline z-theory/generate_z_events.cpp:37]: Writing the entire pythia "event" to the TTree is probably not manageable for any serious number of events.

You can try writing single precision (32 bit) floats for the relevant variables like the mup_PT,mup_ETA,mup_PHI etc...
Something like
```cpp
float mup_PT;
T->Branch("mup_PT",&mup_PT);
for (int iEvent = 0;....){
mup_PT = 1.0;
T->Fill();
}
```

Then you need to figure out how to actually find the muons from Z in pythia....

Comment (mvesteri) [reply z-theory/generate_z_events.cpp:37]: We could try to find the "Z" particle from the event history but I'd first suggest to try and find the two highest pT final state muons (of opposite charge).
Inside each event you can do
```cpp
const auto & event = pythia.event ;
 for ( auto i_particle = 0; i_particle < event.size(); ++i_particle ) {
    auto pT = event[i_particle].pT(); 
 auto ID = event[i_particle].pid(); 
}
```
Then you should be able to write some logic to find the highest pT final state mu+ and mu-.
Or more useful is to find their indices.
Then you can just ask for whatever property for the particles at those indices and copy their values to the TTree output variables.

**Comment by r-preston** [comment]
>  @r-preston . One quick question. Can you see the following file?
Yep, can read it

**Comment by mvesteri** [comment]
OK not really critical right now since your code and `Makefile` are in a different directory to Laura's code, but I think we could play around with using the precompiled pythia library and header from `cvfms` (which is the CERN way of distributing a lot of software in a dynamic/efficient way). Then we can put both sets of code under the same `Makefile`.

**Comment by r-preston** [comment]
> @r-preston I think you forgot to attach the figures?

Ah sorry that didn't work out how i'd intended, the figure is part of the latest commit because github wouldn't let me put it in a comment; however it shifted my first comment from the commit into the merge request

**Comment by mvesteri** [comment]
Can you not just take a screen shot of the figure and paste to a discussion box here?

**Comment by r-preston** [comment]
2020-11-12-113122_1442x1000_scrot
I didn't realise you can just drag and drop. Sorry about that

**Comment by mvesteri** [comment]
@r-preston as I said in the meeting I think the plot of the pT may be OK if you are producing Z/gamma* down to lower masses. So there will be two populations in the pT distribution. The broad structure around mZ/2 comes from the Z peak. The lower pT muons come from the lower mass part. Suggest asking pythia to generate between 60 and 120 GeV. It is more efficient to tell pythia to do this, rather than just applying the cut yourself. Otherwise pythia will waste a lot of CPU [randomly / Monte Carlo] sampling the low mass region.

**Comment by r-preston** [comment]
@mvesteri @lmcairns Code is producing consistent values, not sure if they're the right values though!
2020-11-26-165645_463x212_scrot

**Comment by mvesteri** [comment]
👍  @r-preston  are you sure that the sqrts value is 5 TeV? The cross section seems a bit high?

This is for  '60 < M < 120 GeV AND both muons with pT > 20 GeV and 2 < eta < 4.5?

**Comment by r-preston** [comment]
>  @r-preston are you sure that the sqrts value is 5 TeV? The cross section seems a bit high?

@mvesteri I thought it seemed much too high. Yes, E_cm is 5 TeV

> This is for '60 < M < 120 GeV AND both muons with pT > 20 GeV and 2 < eta < 4.5?

What the code currently does is export the highest energy muon pair for each event along with global data about cross section, then the python code works out the cross section for a mass cut, muon momentum cut and eta cut by weighting each muon as total cross section/number of events.

**Pull request merged**